File: ruby_dsa/lists/singly/impl.rb
class SinglyNode
  attr_accessor :value, :next_node

  def initialize(value)
    @value = value
    @next_node = nil
  end
end

class Singly
  attr_reader :head, :tail, :size

  def initialize
    @head = nil
    @tail = nil
    @size = 0
  end

  def push(value)
    new_node = SinglyNode.new(value)

    if empty?
      @head = new_node
      @tail = @head
    else
      @tail.next_node = new_node
      @tail = new_node
    end

    @size += 1
  end

  def pop
    return nil if empty?

    if head == tail
      @head = nil
      @tail = nil
    else
      current = @head

      current = current.next_node while current.next_node != @tail

      current.next_node = nil
      @tail = current
    end

    @size -= 1
  end

  def search(value)
    return nil if empty?

    current = @head

    while current
      return current if value == current.value

      current = current.next_node
    end
  end

  def invert
    prev = nil
    current = @head

    while current
      temp = current.next_node
      current.next_node = prev
      prev = current
      current = temp
    end

    @head, @tail = @tail, @head
  end

  def empty?
    head.nil? && tail.nil? && size.zero?
  end
end

File: python_dsa/lists/singly/impl.py
from logging import currentframe


class SinglyNode:
    def __init__(self, value):
        self.value = value
        self.next = None


class Singly:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def push(self, value):
        new_node = SinglyNode(value)

        if self.is_empty():
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

        self.size += 1

    def pop(self):
        if self.is_empty():
            return None

        current = self.head

        while current.next != self.tail:
            current = current.next

        current.next = None
        self.tail = current

        self.size -= 1

    def search(self, value):
        if self.is_empty():
            return None

        current = self.head

        while current:
            if current.value == value:
                return current
            current = current.next

    def invert(self):
        prev = None
        current = self.head

        while current:
            temp = current.next
            current.next = prev
            prev = current
            current = temp

        self.head, self.tail = self.tail, self.head

    def is_empty(self):
        return self.head is None and self.tail is None and self.size == 0
